c#/Traverse/TraverseFolders/Program.cs

```diff
@@ -6,15 +6,68 @@ class Program
     {
         static void Main(string[] args)
         {
-            var fileSystemVisitor1 = new FileSystemVisitor("../../../../testFolder");
+            /*
+            var visitor = new FileSystemVisitor("../../../../testFolder");
             
-            ReturnedDelegate filter = delegate(string filename, string extention)
+            Console.WriteLine("===== Result from first visitor ======");
+            foreach (var file in visitor.Explore("../../../../testFolder"))
             {
-                string[] newStr = filename.Split("/");
-                string[] newstring = newStr[newStr.Length -1].Split(".");
-                return newstring[newstring.Length -1] == extention;
-            };
-            var fileSystemVisitor2 = new FileSystemVisitor("../../../../testFolder", filter);
+                Console.WriteLine(file);
+            }
+            
+            var visitor2 = new FileSystemVisitor("../../../../testFolder",FilterByCsExtension);
+            
+            Console.WriteLine("===== Result from second visitor ======");
+            
+            foreach (var file in visitor2.Explore("../../../../testFolder"))
+            {
+                Console.WriteLine(file);
+            }
+            */
+            var visitor3 = new FileSystemVisitor("../../../../testFolder",FilterByCsExtension,ShowStartMessage,ShowFinishMessage,FileFoundEvent);
+            
+            foreach (var file in visitor3.Explore("../../../../testFolder"))
+            {
+                Console.WriteLine(file);
+            }
+        }
+
+        static bool FileFoundEvent(string file,CustomEventArgs args)
+        {
+            var abortionFile = "../../../../testFolder/file1.cs";
+            var e = new CustomEventArgs();
```

Comment:
you don't need to create new CustomEventArgs here, if you modify "args" from parameters here, this change will be visible outside this method. For the same reason you don't need to return boolean from this method

Reply:
the rest of the logic here is correct, similar logic I described above and the same way you can implement all other events